### ISSUE TITLE
Destination array was not long enough. Check the destination index, length, and the array's lower bounds

### ISSUE THREAD

**Issue by igortech** [linq2db/linq2db#3056]
```
Destination array was not long enough. Check the destination index, length, and the array's lower bounds. 

   at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)\r\n   at LinqToDB.Metadata.MetadataReader.GetDynamicColumns(Type type)\r\n   at LinqToDB.Mapping.MappingSchema.GetDynamicColumns(Type type)\r\n   at LinqToDB.Mapping.EntityDescriptor.Init()\r\n   at LinqToDB.Mapping.EntityDescriptor..ctor(MappingSchema mappingSchema, Type type)\r\n   at LinqToDB.Mapping.MappingSchema.<>c.<GetEntityDescriptor>b__94_0(ICacheEntry`1 o, MappingSchema context)\r\n   at LinqToDB.Common.Internal.Cache.CacheExtensions.GetOrCreate[TItem,TKey,TContext](IMemoryCache`1 cache, TKey key, TContext context, Func`3 factory)\r\n   at LinqToDB.Mapping.MappingSchema.GetEntityDescriptor(Type type)\r\n   at LinqToDB.Linq.Table`1.InitTable(IDataContext dataContext, Expression expression)\r\n   at LinqToDB.Linq.Table`1..ctor(IDataContext dataContext)\r\n   at LinqToDB.Data.DataConnection.GetTable[T]()\r\n   at LinqToDB.Data.DataConnectionExtensions.BulkCopy[T](DataConnection dataConnection, BulkCopyOptions options, 
```

### Steps to reproduce
in version 3.3 - it's working
in version 3.4 - error

```c#
    [TestClass]
    public class DataModelTest
    {
       [TestMethod]
        public void DataModelDynamicTableTest2()
        {
            using (var dm = new DataModel("SqlServer", "Data Source=.;Initial Catalog=testdb;Integrated Security=sspi;"))
            {
                dm.CommandTimeout = 50;
                dm.DropTable<TestRow>(throwExceptionIfNotExists: false);
                dm.CreateTable<TestRow>();

                var fm = MappingSchema.Default.GetFluentMappingBuilder();

                fm.Entity<DynamicTableRow>()
                    //.HasIdentity(x => Sql.Property<int>(x, "Id"))
                    .Property(x => Sql.Property<string>(x, "Name"))
                    .Property(x => Sql.Property<string>(x, "Description"));

                var rows = new List<DynamicTableRow>();
                var drow = new DynamicTableRow();
                drow.Properties.Add("Id", 1);
                drow.Properties.Add("Name", "n1");
                drow.Properties.Add("Description", "d0");
                rows.Add(drow);
                drow = new DynamicTableRow();
                drow.Properties.Add("Id", 2);
                drow.Properties.Add("Name", "n2");
                drow.Properties.Add("Description", "d00");
                rows.Add(drow);

                dm.BulkCopy(new BulkCopyOptions
                {
                    TableName = "TestRow",
                    SchemaName = "dbo",
                    MaxBatchSize = 100
                },
                 rows);
            }
        }
    }

    [Table("TestRow", Schema = "dbo")]
    class TestRow
    {
        [PrimaryKey, Identity]
        public int Id;

        [Column(DataType = DataType.VarChar, Length = 100)]
        public string Name;

        [Column(DataType = DataType.VarChar, Length = 500)]
        public string Description;

    }

    [Table]
    public class DynamicTableRow
    {
        [DynamicColumnsStore]
        public Dictionary<string, object> Properties = new Dictionary<string, object>();
    }
```

### Environment details
linq2db version: 3.4.0
Database Server: SqlServer 2014
Database Provider: SqlServer
Operating system: Windows 10
.NET Framework: core 3.1

**Comment by to11mtm** [comment]
Tagging @MaceWindu for awareness, I think this was a result of #2971,

In [MetadataReader.cs](https://github.com/linq2db/linq2db/blob/43ca3371c2f7b0c99305d8e427885a52c5ee4a9d/Source/LinqToDB/Metadata/MetadataReader.cs#L124-L133) (comments mine)
```
			length = 0;
			var columns = new MemberInfo[length]; //length was reset, so this is always 0
			for (var i = 0; i < cols.Length; i++)
			{
				if (cols[i].Length > 0)
				{
					Array.Copy(cols[i], 0, columns, length, cols[i].Length);
					length += cols[i].Length;
				}
			}
```

Should be an easy fix, I'll work on a PR :)

**Comment by to11mtm** [comment]
Re-Opening until 3.4.1 is released

**Comment by MaceWindu** [comment]
@to11mtm we don't keep issues open for merged fixes, so I would prefer to stick to current workflow. It doesn't overload milestone with open tasks that should be actually closed.

**Comment by to11mtm** [comment]
My apologies! will re-close. @igortech  this should be fixed in the upcoming point release (3.4.1)
